echopype/preprocess/api.py

```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+    grps_same_ind = [
+        np.argwhere(unique_inverse == grp).flatten() for grp in np.unique(unique_inverse)
+    ]
+
+    # for each group bin and mean arr along echo_range
+    # note: the values appended may not be in the correct final order
+    binned_er = []
+    for count, grp in enumerate(grps_same_ind):
+        binned_er.append(
+            bin_and_mean_echo_range(arr[grp, :], unique_er_bin_ind[count, :], n_bin_er)
+        )
+
+    # construct er_means and put the columns in the correct order
+    binned_er_array = np.hstack(binned_er)
+    correct_column_ind = np.argsort(np.concatenate(grps_same_ind))
+    er_means = binned_er_array[:, correct_column_ind]
+
+    return er_means
+
+
+def bin_and_mean_2d(
+    arr: Union[dask.array.Array, np.ndarray],
+    bins_time: np.ndarray,
+    bins_er: np.ndarray,
+    times: np.ndarray,
+    echo_range: np.ndarray,
+    comp_er_check: bool = True,
+) -> np.ndarray:
+    """
+    Bins and means ``arr`` based on ``times`` and ``echo_range``,
+    and their corresponding bins. If ``arr`` is ``Sv`` then this
+    will compute the MVBS.
+
+    Parameters
+    ----------
+    arr: dask.array.Array or np.ndarray
+        The 2D array whose values should be binned
+    bins_time: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``times``
+    bins_er: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``echo_range``
+    times: np.ndarray
+        1D array corresponding to the time values that should be binned
+    echo_range: np.ndarray
+        2D array of echo range values
+    comp_er_check: bool
+        If True, a more comprehensive check will be completed to determine if ``echo_range``
+        grouping along ``ping_time`` is needed, otherwise a less comprehensive check will be done
+
+    Returns
+    -------
+    final_reduced: np.ndarray
+        The final binned and mean ``arr``, if ``arr`` is ``Sv`` then this is the MVBS
+
+    Notes
+    -----
+    This function assumes that ``arr`` has rows corresponding to
+    ``ping_time`` and columns corresponding to ``echo_range``.
+
+    This function should not be run if the number of ``echo_range`` values
+    vary amongst ``ping_times``.
+    """
+
+    # get the number of echo range and time bins
+    n_bin_er = len(bins_er)
+    n_bin_time = len(bins_time)
+
+    # obtain the bin indices for echo_range and times
+    digitized_echo_range, bin_time_ind = get_bin_indices(echo_range, bins_er, times, bins_time)
+
+    # determine if grouping of echo_range values with the same step size is necessary
+    if comp_er_check:
+        grouping_needed = is_grouping_needed_comprehensive(echo_range)
+    else:
+        grouping_needed = is_grouping_needed_less_comprehensive(echo_range)
+
+    if grouping_needed:
+        # groups, bins, and means arr with respect to echo_range
+        er_means = group_bin_mean_echo_range(arr, digitized_echo_range, n_bin_er)
```

Comment:
For the sake of clarify, maybe change this function name to `er_group_bin_mean_echo_range`

Reply:
Why did you add `er`? The way I read `group_bin_mean_echo_range` is: group first, bin these values, and then take the mean along `echo_range`.  If you are trying to specify that the grouping is with respect to `digitized_echo_range`, then I suggest the following name `group_dig_er_bin_mean_echo_range`.

Reply:
I like `group_dig_er_bin_mean_echo_range` (even though one can say it's long :P) -- wanted to make this more specific just because it is not clear what is being grouped, since binning is already a type of grouping.

Reply:
Yes, I agree that it is a bit long, but it describes very specifically what is happening. I will go ahead and name it `group_dig_er_bin_mean_echo_range`, since you seem to be fine with it.